File: ingest/classifier.py
"""classifier.py — Tier-1 ФИЛЬТР чувствительности + роутер «задача vs знание».

НАЗНАЧЕНИЕ ([ADR-0011])
======================
Sibling к `sanitizer.py`, НЕ внутри него — противоположная семантика отказа:
  • sanitizer: секреты fail-CLOSED (abort write), PII mask-in-place.
  • classifier: чувствительность fail-TO-QUARANTINE (маршрутизирует ЦЕЛЫЙ документ),
    важность/лейн — выбор куда положить, без удаления.

Две оси + роутер (никогда не один гейт «выкинуть плохое/ненужное»):
  • Ось A (чувствительность) — здесь, on-device, ДО облака. Tier-1 детерминированный:
    source_class + domain_blocklist + (opt-in) lexicon + pii_density. БЕЗ ML, БЕЗ embedder.
  • Ось B (важность) — НЕ здесь: расширение compile-решения (LLM 1–10 как тай-брейкер).
  • Роутер «задача vs знание» — здесь, консервативен В СТОРОНУ ЗНАНИЯ.

ИНВАРИАНТЫ
  • Finding-discipline: возвращаем label/score/tier/reason/action — НИКОГДА не содержимое.
  • Карантин предпочитает ложно-ПОЛОЖИТЕЛЬНЫЕ (raw/ хранит → дёшево); drop-from-wiki — нет.
  • Карантин ПОБЕЖДАЕТ лейн (чувствительная чора → карантин, не .tasks/).
  • Политика — из compiler/relevance-policy.md (JSON-блок) + приватный .local.json override.
  • stdlib-only. Tier-2 ML — ОТЛОЖЕН (см. ADR-0011).
"""
from __future__ import annotations

import hashlib
import json
import re
import unicodedata
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path

# Нормализацию переиспользуем из sanitizer, чтобы обфускация не проскочила лексикон.
try:
    from .sanitizer import _strip_invisible as _san_strip  # type: ignore
except Exception:  # pragma: no cover - sanitizer может быть импортирован иначе
    _san_strip = None

# --- Действия (по убыванию строгости: меньший индекс = строже = ПОБЕЖДАЕТ) ------
_ACTION_PRECEDENCE = [
    "quarantine",            # целый док в raw/.quarantine/<cat>/
    "quarantine_and_redact",
    "keep_redact_spans",     # ХРАНИМ как знание; sanitizer маскирует опасные подстроки
    "leave_in_raw",          # не промоутить в wiki/ (но raw/ хранит)
    "normal",                # обычный путь
]
def _stricter(a: str, b: str) -> str:
    ia = _ACTION_PRECEDENCE.index(a) if a in _ACTION_PRECEDENCE else 99
    ib = _ACTION_PRECEDENCE.index(b) if b in _ACTION_PRECEDENCE else 99
    return a if ia <= ib else b

# Безопасные ОБЩИЕ дефолт-кейворды (расширение — в приватном .local.json).
_DEFAULT_KEYWORDS = {
    "financial": ["долг", "должен", "займ", "iban", "оплат", "счёт", "invoice", "owe", "debt"],
    "health":    ["диагноз", "болезн", "симптом", "therapy", "терап", "health", "лекарств"],
    "legal":     ["контракт", "договор", "иск", "суд", "nda", "lawsuit", "арбитраж"],
    "toxic":     [],   # реальный лексикон — приватно; по умолчанию пусто (RU informal даёт ложные)
}
_EMAIL_RE = re.compile(r"[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}")
_PHONE_RE = re.compile(r"(?<!\d)(?:\+?\d[\d\-\s()]{8,}\d)(?!\d)")


@dataclass(frozen=True)
class Classification:
    """Результат оси A. Содержит только метаданные — НИКОГДА не текст."""
    label: str          # nsfw | others_pii | toxic | financial | health | legal | doomscroll | normal
    action: str         # см. _ACTION_PRECEDENCE
    tier: int           # 1 = детерминированный on-device
    reason: str         # человекочитаемая причина (provenance), без содержимого
    score: float | None = None


@dataclass(frozen=True)
class LaneDecision:
    lane: str           # "knowledge" | "task"
    dual_route: bool    # True → и в task-log, И видимо для compile (на сомнении)
    reason: str


# --- Загрузка политики --------------------------------------------------------
_JSON_BLOCK = re.compile(r"```json\s*(\{.*?\})\s*```", re.S)

def load_policy(public_md: str | Path | None = None,
                private_override: str | Path | None = None) -> dict:
    """Читает JSON-блок из compiler/relevance-policy.md + мёрджит приватный .local.json.

    Граница двух репо: персональные лексиконы/имена живут ТОЛЬКО в приватном override.
    """
    md = Path(public_md) if public_md else Path(__file__).resolve().parent.parent / "compiler" / "relevance-policy.md"
    policy: dict = {}
    try:
        m = _JSON_BLOCK.search(md.read_text(encoding="utf-8"))
        if m:
            policy = json.loads(m.group(1))
    except Exception:
        policy = {}
    if private_override and Path(private_override).exists():
        try:
            _deep_merge(policy, json.loads(Path(private_override).read_text(encoding="utf-8")))
        except Exception:
            pass
    return policy

def _deep_merge(base: dict, over: dict) -> dict:
    for k, v in over.items():
        if isinstance(v, dict) and isinstance(base.get(k), dict):
            _deep_merge(base[k], v)
        else:
            base[k] = v
    return base


def _normalize(text: str) -> str:
    text = unicodedata.normalize("NFKC", text)
    if _san_strip:
        try:
            text = _san_strip(text)
        except Exception:
            pass
    return text.lower()


# --- Детекторы (Tier-1, детерминированные) ------------------------------------
def _match_source_class(meta: dict, classes: list[str]) -> bool:
    sc = (meta or {}).get("source_class", "")
    return bool(sc) and sc in set(classes or [])

def _match_domain_blocklist(text: str, meta: dict, blocklist: list[str]) -> str | None:
    hay = " ".join([text, " ".join(str(v) for v in (meta or {}).values())]).lower()
    for dom in (blocklist or []):
        d = str(dom).split("#")[0].strip().lower()
        if d and d in hay:
            return d
    return None

def _match_keyword(text: str, words: list[str]) -> str | None:
    for w in (words or []):
        w = w.strip().lower()
        if w and w in text:
            return w
    return None

def _pii_density(text: str) -> int:
    return len(_EMAIL_RE.findall(text)) + len(_PHONE_RE.findall(text))


# --- Ось A: классификация чувствительности ------------------------------------
def classify_sensitivity(text: str, source_meta: dict | None = None,
                         policy: dict | None = None) -> Classification:
    """Tier-1 детерминированная классификация ЦЕЛОГО документа. On-device, без egress."""
    policy = policy or load_policy()
    meta = source_meta or {}
    norm = _normalize(text or "")
    sens = (policy.get("sensitivity") or {})

    best = Classification(label="normal", action="normal", tier=1, reason="no-signal")
    for cat, cfg in sens.items():
        cfg = cfg or {}
        detect = cfg.get("detect", [])
        action = cfg.get("action", "normal")
        hit_reason = None

        if "source_class" in detect and _match_source_class(meta, cfg.get("source_classes", _SRC_DEFAULTS.get(cat, []))):
            hit_reason = f"source_class={meta.get('source_class')}"
        if not hit_reason and "domain_blocklist" in detect:
            dom = _match_domain_blocklist(norm, meta, cfg.get("domain_blocklist", []))
            if dom:
                hit_reason = f"domain_blocklist={dom}"
        if not hit_reason and "keyword" in detect:
            kw = _match_keyword(norm, cfg.get("keywords", _DEFAULT_KEYWORDS.get(cat, [])))
            if kw:
                hit_reason = f"keyword={kw}"
        if not hit_reason and "lexicon" in detect and cfg.get("lexicon") == "on":
            kw = _match_keyword(norm, cfg.get("lexicon_words", []))
            if kw:
                hit_reason = f"lexicon={kw}"
        if not hit_reason and "pii_density" in detect and _pii_density(norm) >= int(cfg.get("pii_density_threshold", 5)):
            hit_reason = f"pii_density>={cfg.get('pii_density_threshold', 5)}"

        if hit_reason:
            cand = Classification(label=cat, action=action, tier=1, reason=hit_reason)
            # Карантин ПОБЕЖДАЕТ: берём более строгое действие.
            if _stricter(cand.action, best.action) == cand.action and cand.action != best.action:
                best = cand
            elif best.label == "normal":
                best = cand
    return best


# дефолтные source_class-метки по категориям (расширяй в политике/приватно)
_SRC_DEFAULTS = {
    "nsfw": ["adult", "porn"],
    "doomscroll": ["feed", "shorts", "reels"],
}


# --- Роутер «задача vs знание» (консервативен В СТОРОНУ ЗНАНИЯ) ----------------
def route_lane(text: str, source_meta: dict | None = None,
               policy: dict | None = None) -> LaneDecision:
    """Дивертит в TASK только при форме «императив + объект». На сомнении — дуал-роут."""
    policy = policy or load_policy()
    lanes = (policy.get("lanes") or {})
    if lanes.get("router_bias", "knowledge") != "knowledge":
        pass  # сейчас поддерживаем только knowledge-bias; иное оставляем knowledge
    norm = _normalize(text or "")
    triggers = [t.lower() for t in lanes.get("task_triggers", [])]
    hit = next((t for t in triggers if t in norm), None)
    if not hit:
        return LaneDecision(lane="knowledge", dual_route=False, reason="no-task-trigger")

    # «императив + объект»: после триггера есть содержательный объект, и это не рефлексия.
    looks_imperative = bool(re.search(r"(^|[\.\!\?]\s*)(" + re.escape(hit) + r")\b", norm)) or norm.strip().startswith(hit)
    has_object = len(norm.split()) >= 2
    reflection_markers = ("я думаю", "хочу научиться", "по жизни", "в целом", "delegat", "career", "карьер")
    looks_reflection = any(m in norm for m in reflection_markers)

    if looks_imperative and has_object and not looks_reflection:
        return LaneDecision(lane="task", dual_route=bool(lanes.get("dual_route_on_ambiguity", True)),
                            reason=f"task_trigger={hit};imperative")
    # Совпало слово, но форма неоднозначна → дуал-роут (и task, и видимо compile), не теряем знание.
    return LaneDecision(lane="knowledge",
                        dual_route=bool(lanes.get("dual_route_on_ambiguity", True)),
                        reason=f"ambiguous_trigger={hit};kept-as-knowledge")


# --- P0-1: явное исключение dot-папок (.quarantine/.tasks/.watermarks) ---------
def should_skip_raw_path(path: str | Path) -> bool:
    """ЛЮБОЙ читатель raw/ обязан вызывать это: rglob НЕ пропускает dot-папки сам.

    Без явной проверки .quarantine/ и .tasks/ снова попадут в compile/query/digest
    → ломается изоляция карантина (P0-1 из adversarial-проверки ADR-0011).
    """
    return any(part.startswith(".") for part in Path(path).parts)


# --- Аудит-лог: строка ledger (НИКОГДА не содержимое) --------------------------
def filter_log_record(raw_path: str, clf: Classification, *, axis: str,
                      lane: LaneDecision | None = None,
                      content: str | None = None, policy_version: str = "") -> dict:
    rec = {
        "ts": datetime.now(timezone.utc).isoformat(),
        "raw_path": raw_path,
        "axis": axis,
        "category": clf.label,
        "action": clf.action,
        "tier": clf.tier,
        "reason": clf.reason,
        "score": clf.score,
        "policy_version": policy_version,
    }
    if content is not None:
        rec["content_sha256"] = "sha256:" + hashlib.sha256(content.encode("utf-8", "replace")).hexdigest()
    if lane is not None:
        rec["lane"] = lane.lane
        rec["dual_route"] = lane.dual_route
    return rec


if __name__ == "__main__":  # быстрый дым-тест на СИНТЕТИКЕ
    pol = load_policy()
    for t, meta in [
        ("Я должен Ивану 5000 за билеты", {"source_class": "telegram"}),
        ("купи молоко и хлеб", {"source_class": "telegram"}),
        ("хочу научиться делегировать, чтобы buy back my time", {"source_class": "note"}),
        ("explicit adult content", {"source_class": "adult"}),
    ]:
        c = classify_sensitivity(t, meta, pol)
        l = route_lane(t, meta, pol)
        print(f"sens={c.label}/{c.action} ({c.reason}) | lane={l.lane} dual={l.dual_route} ({l.reason})")

File: src/ingest/classifier.ts
/**
 * classifier.ts — Tier-1 ФИЛЬТР чувствительности + роутер «задача vs знание».
 *
 * Порт `ingest/classifier.py` ([ADR-0011], [ADR-0012]). Контракт сохранён:
 *   classifySensitivity / routeLane / shouldSkipRawPath / filterLogRecord / loadPolicy
 *
 * Sibling к sanitizer (НЕ внутри — противоположная семантика отказа): секреты
 * fail-CLOSED, чувствительность fail-TO-QUARANTINE. Две оси + роутер:
 *  - Ось A (чувствительность) — здесь, on-device, ДО облака. Tier-1 детерминированный.
 *  - Ось B (важность) — НЕ здесь (расширение compile).
 *  - Роутер «задача vs знание» — здесь, консервативен В СТОРОНУ ЗНАНИЯ.
 *
 * Инварианты: возвращаем label/score/tier/reason/action — НИКОГДА содержимое;
 * карантин предпочитает ложно-ПОЛОЖИТЕЛЬНЫЕ; карантин ПОБЕЖДАЕТ лейн; политика —
 * из compiler/relevance-policy.md (JSON-блок) + приватный .local.json override.
 */
import { createHash } from 'node:crypto';
import { existsSync, readFileSync } from 'node:fs';
import { resolve } from 'node:path';

import { looksLikePhone, stripInvisible } from './sanitizer.js';

// --- Действия (по убыванию строгости: меньший индекс = строже = ПОБЕЖДАЕТ) ----
const ACTION_PRECEDENCE = [
	'quarantine', // целый док в raw/.quarantine/<cat>/
	'quarantine_and_redact',
	'keep_redact_spans', // ХРАНИМ как знание; sanitizer маскирует опасные подстроки
	'leave_in_raw', // не промоутить в wiki/ (но raw/ хранит)
	'normal', // обычный путь
];

function stricter(a: string, b: string): string {
	const ia = ACTION_PRECEDENCE.indexOf(a);
	const ib = ACTION_PRECEDENCE.indexOf(b);
	const sa = ia < 0 ? 99 : ia;
	const sb = ib < 0 ? 99 : ib;
	return sa <= sb ? a : b;
}

// Безопасные ОБЩИЕ дефолт-кейворды (расширение — в приватном .local.json).
const DEFAULT_KEYWORDS: Record<string, string[]> = {
	financial: ['долг', 'должен', 'займ', 'iban', 'оплат', 'счёт', 'invoice', 'owe', 'debt'],
	health: ['диагноз', 'болезн', 'симптом', 'therapy', 'терап', 'health', 'лекарств'],
	legal: ['контракт', 'договор', 'иск', 'суд', 'nda', 'lawsuit', 'арбитраж'],
	toxic: [], // реальный лексикон — приватно
};

// Дефолтные source_class-метки по категориям (расширяй в политике/приватно).
const SRC_DEFAULTS: Record<string, string[]> = {
	nsfw: ['adult', 'porn'],
	doomscroll: ['feed', 'shorts', 'reels'],
};

// Маркеры рефлексии — роутер НЕ дивертит их в task (это ростовой сигнал).
const REFLECTION_MARKERS = [
	'я думаю',
	'хочу научиться',
	'по жизни',
	'в целом',
	'delegat',
	'career',
	'карьер',
];

const EMAIL_RE = /[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}/g;
const PHONE_RE = /(?<!\d)\+?\d[\d\-\s()]{8,}\d(?!\d)/g;

// --- Типы ---------------------------------------------------------------------
export interface Classification {
	label: string; // nsfw | others_pii | toxic | financial | health | legal | doomscroll | normal
	action: string; // см. ACTION_PRECEDENCE
	tier: number; // 1 = детерминированный on-device
	reason: string; // человекочитаемая причина (provenance), без содержимого
	score: number | null;
}

export interface LaneDecision {
	lane: 'knowledge' | 'task';
	dualRoute: boolean; // true → и в task-log, И видимо для compile (на сомнении)
	reason: string;
}

export interface SourceMeta {
	source_class?: string;
	[k: string]: unknown;
}

interface SensitivityCfg {
	detect?: string[];
	action?: string;
	source_classes?: string[];
	domain_blocklist?: string[];
	keywords?: string[];
	lexicon?: string;
	lexicon_words?: string[];
	pii_density_threshold?: number;
	[k: string]: unknown;
}

interface LanesCfg {
	router_bias?: string;
	task_triggers?: string[];
	dual_route_on_ambiguity?: boolean;
	[k: string]: unknown;
}

export interface Policy {
	sensitivity?: Record<string, SensitivityCfg>;
	lanes?: LanesCfg;
	[k: string]: unknown;
}

export interface FilterLogRecord {
	ts: string;
	raw_path: string;
	axis: string;
	category: string;
	action: string;
	tier: number;
	reason: string;
	score: number | null;
	policy_version: string;
	content_sha256?: string;
	lane?: string;
	dual_route?: boolean;
}

// --- Загрузка политики --------------------------------------------------------
// Из src/ingest/classifier.ts и dist/ingest/classifier.js одинаково: ../../compiler.
const DEFAULT_POLICY_PATH = resolve(
	import.meta.dirname,
	'..',
	'..',
	'compiler',
	'relevance-policy.md',
);
const JSON_BLOCK_RE = /```json\s*(\{[\s\S]*?\})\s*```/;

function isPlainObject(v: unknown): v is Record<string, unknown> {
	return typeof v === 'object' && v !== null && !Array.isArray(v);
}

function deepMerge(base: Record<string, unknown>, over: Record<string, unknown>): Record<string, unknown> {
	for (const [k, v] of Object.entries(over)) {
		const cur = base[k];
		if (isPlainObject(v) && isPlainObject(cur)) {
			deepMerge(cur, v);
		} else {
			base[k] = v;
		}
	}
	return base;
}

/**
 * Читает JSON-блок из compiler/relevance-policy.md + мёрджит приватный .local.json.
 * Граница двух репо: персональные лексиконы/имена живут ТОЛЬКО в приватном override.
 */
export function loadPolicy(publicMd?: string | null, privateOverride?: string | null): Policy {
	const mdPath = publicMd ?? DEFAULT_POLICY_PATH;
	let policy: Policy = {};
	try {
		const text = readFileSync(mdPath, 'utf8');
		const m = JSON_BLOCK_RE.exec(text);
		if (m && m[1]) policy = JSON.parse(m[1]) as Policy;
	} catch {
		policy = {};
	}
	if (privateOverride && existsSync(privateOverride)) {
		try {
			deepMerge(policy as Record<string, unknown>, JSON.parse(readFileSync(privateOverride, 'utf8')));
		} catch {
			// приватный override битый — игнор (как в Python)
		}
	}
	return policy;
}

// Нормализацию переиспользуем из sanitizer, чтобы обфускация не проскочила лексикон.
function normalize(text: string): string {
	return stripInvisible(text).toLowerCase();
}

// --- Детекторы (Tier-1, детерминированные) ------------------------------------
function matchSourceClass(meta: SourceMeta, classes: string[]): boolean {
	const sc = meta.source_class ?? '';
	return Boolean(sc) && (classes ?? []).includes(sc);
}

function matchDomainBlocklist(text: string, meta: SourceMeta, blocklist: string[]): string | null {
	const metaVals = Object.values(meta)
		.map((v) => String(v))
		.join(' ');
	const hay = `${text} ${metaVals}`.toLowerCase();
	for (const dom of blocklist ?? []) {
		const d = (String(dom).split('#')[0] ?? '').trim().toLowerCase();
		if (d && hay.includes(d)) return d;
	}
	return null;
}

function matchKeyword(text: string, words: string[]): string | null {
	for (const raw of words ?? []) {
		const w = raw.trim().toLowerCase();
		if (w && text.includes(w)) return w;
	}
	return null;
}

function piiDensity(text: string): number {
	const emails = text.match(EMAIL_RE)?.length ?? 0;
	// Считаем только ВАЛИДНЫЕ телефоны (10..15 значащих цифр, как в sanitizer-ярус-2).
	// Иначе phone-regex ловит фрагменты ISO-дат (2024-05-29 = 8 цифр), и любая
	// дат-насыщенная страница (frontmatter + per-message stamps) ложно перебирает
	// порог → others_pii-карантин. ОСОЗНАННОЕ расхождение с Python-оригиналом
	// (фикс пред-существующего бага, НЕ редизайн фильтра) — [ADR-0013].
	const phones = (text.match(PHONE_RE) ?? []).filter((m) => looksLikePhone(m)).length;
	return emails + phones;
}

// --- Ось A: классификация чувствительности ------------------------------------
/** Tier-1 детерминированная классификация ЦЕЛОГО документа. On-device, без egress. */
export function classifySensitivity(
	text: string,
	sourceMeta?: SourceMeta | null,
	policy?: Policy | null,
): Classification {
	const pol = policy ?? loadPolicy();
	const meta = sourceMeta ?? {};
	const norm = normalize(text ?? '');
	const sens: Record<string, SensitivityCfg> = pol.sensitivity ?? {};

	let best: Classification = {
		label: 'normal',
		action: 'normal',
		tier: 1,
		reason: 'no-signal',
		score: null,
	};

	for (const [cat, rawCfg] of Object.entries(sens)) {
		const cfg = rawCfg ?? {};
		const detect = cfg.detect ?? [];
		const action = cfg.action ?? 'normal';
		let hitReason: string | null = null;

		if (
			detect.includes('source_class') &&
			matchSourceClass(meta, cfg.source_classes ?? SRC_DEFAULTS[cat] ?? [])
		) {
			hitReason = `source_class=${meta.source_class ?? ''}`;
		}
		if (!hitReason && detect.includes('domain_blocklist')) {
			const dom = matchDomainBlocklist(norm, meta, cfg.domain_blocklist ?? []);
			if (dom) hitReason = `domain_blocklist=${dom}`;
		}
		if (!hitReason && detect.includes('keyword')) {
			const kw = matchKeyword(norm, cfg.keywords ?? DEFAULT_KEYWORDS[cat] ?? []);
			if (kw) hitReason = `keyword=${kw}`;
		}
		if (!hitReason && detect.includes('lexicon') && cfg.lexicon === 'on') {
			const kw = matchKeyword(norm, cfg.lexicon_words ?? []);
			if (kw) hitReason = `lexicon=${kw}`;
		}
		if (!hitReason && detect.includes('pii_density')) {
			// Math.trunc + finite-guard: усечение дробного порога как Python int()
			// (5.9 → 5, без off-by-one на границе карантина) и без NaN-fail-open на
			// нечисловом значении (тогда сигнал просто не срабатывает) — [аудит TS-порта].
			const threshold = Math.trunc(Number(cfg.pii_density_threshold ?? 5));
			if (Number.isFinite(threshold) && piiDensity(norm) >= threshold) {
				hitReason = `pii_density>=${threshold}`;
			}
		}

		if (hitReason) {
			const cand: Classification = { label: cat, action, tier: 1, reason: hitReason, score: null };
			// Карантин ПОБЕЖДАЕТ: берём более строгое действие.
			if (stricter(cand.action, best.action) === cand.action && cand.action !== best.action) {
				best = cand;
			} else if (best.label === 'normal') {
				best = cand;
			}
		}
	}
	return best;
}

// --- Роутер «задача vs знание» (консервативен В СТОРОНУ ЗНАНИЯ) ----------------
/** Дивертит в TASK только при форме «императив + объект». На сомнении — дуал-роут. */
export function routeLane(
	text: string,
	_sourceMeta?: SourceMeta | null,
	policy?: Policy | null,
): LaneDecision {
	const pol = policy ?? loadPolicy();
	const lanes: LanesCfg = pol.lanes ?? {};
	// Сейчас поддерживаем только knowledge-bias; иное оставляем knowledge.
	const norm = normalize(text ?? '');
	const triggers = (lanes.task_triggers ?? []).map((t) => t.toLowerCase());
	const hit = triggers.find((t) => norm.includes(t)) ?? null;
	const dualOnAmbiguity = Boolean(lanes.dual_route_on_ambiguity ?? true);

	if (!hit) {
		return { lane: 'knowledge', dualRoute: false, reason: 'no-task-trigger' };
	}

	// «императив + объект»: триггер в начале или после конца предложения, объект есть,
	// и это не рефлексия. Граница справа — Unicode-aware (Python `\b` юникод-aware,
	// JS `\b` ASCII-only → используем negative-lookahead на букву/цифру/`_`).
	const escaped = hit.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
	const impRe = new RegExp(`(^|[.!?]\\s*)(${escaped})(?![\\p{L}\\p{N}_])`, 'u');
	const looksImperative = impRe.test(norm) || norm.trim().startsWith(hit);
	const hasObject = norm.split(/\s+/).filter((w) => w.length > 0).length >= 2;
	const looksReflection = REFLECTION_MARKERS.some((m) => norm.includes(m));

	if (looksImperative && hasObject && !looksReflection) {
		return { lane: 'task', dualRoute: dualOnAmbiguity, reason: `task_trigger=${hit};imperative` };
	}
	// Совпало слово, но форма неоднозначна → дуал-роут, не теряем знание.
	return {
		lane: 'knowledge',
		dualRoute: dualOnAmbiguity,
		reason: `ambiguous_trigger=${hit};kept-as-knowledge`,
	};
}

// --- P0-1: явное исключение dot-папок (.quarantine/.tasks/.watermarks) ---------
/**
 * ЛЮБОЙ читатель raw/ обязан вызывать это: glob/rglob НЕ пропускает dot-папки сам.
 * Без явной проверки .quarantine/ и .tasks/ снова попадут в compile/query/digest
 * → ломается изоляция карантина (P0-1 из adversarial-проверки [ADR-0011]).
 */
export function shouldSkipRawPath(p: string): boolean {
	// pathlib.parts схлопывает '.' и пустые сегменты; повторяем это поведение.
	const parts = p.split(/[\\/]+/).filter((seg) => seg !== '' && seg !== '.');
	return parts.some((seg) => seg.startsWith('.'));
}

// --- Аудит-лог: строка ledger (НИКОГДА не содержимое) --------------------------
export function filterLogRecord(
	rawPath: string,
	clf: Classification,
	opts: {
		axis: string;
		lane?: LaneDecision | null;
		content?: string | null;
		policyVersion?: string;
		now?: Date;
	},
): FilterLogRecord {
	const { axis, lane = null, content = null, policyVersion = '', now } = opts;
	const rec: FilterLogRecord = {
		ts: (now ?? new Date()).toISOString(),
		raw_path: rawPath,
		axis,
		category: clf.label,
		action: clf.action,
		tier: clf.tier,
		reason: clf.reason,
		score: clf.score,
		policy_version: policyVersion,
	};
	if (content !== null) {
		rec.content_sha256 = 'sha256:' + createHash('sha256').update(content, 'utf8').digest('hex');
	}
	if (lane !== null) {
		rec.lane = lane.lane;
		rec.dual_route = lane.dualRoute;
	}
	return rec;
}
